flake.nix

```diff
@@ -236,39 +236,60 @@
           dontInstall = true;
         };
 
-        # Create a development shell of hls project
-        # See https://github.com/NixOS/nixpkgs/blob/5d4a430472cafada97888cc80672fab255231f57/pkgs/development/haskell-modules/make-package-set.nix#L319
-        mkDevShell = hpkgs:
-          with pkgs;
-          hpkgs.shellFor {
-            doBenchmark = true;
-            packages = p:
-              with builtins;
-              map (name: p.${name}) (attrNames
-                (if hpkgs.ghc.version == "9.0.1" then
-                  removeAttrs hlsSources ghc901Config.disabledPlugins
-                else if hpkgs.ghc.version == "9.2.1" then
-                  removeAttrs hlsSources ghc921Config.disabledPlugins
-                else
-                  hlsSources));
```

Comment:
I think one basic motivation of using nix is that we don't want to wast time building these Haskell dependencies locally. Yes, the build failures happen, and it took a lot of efforts to maintain nix files. The initial idea here was to rely on nix as much as possible, so that we could build only few packages by cabal. Instead of completely cutting down haskell dependencies in nix shell, is there any chance to find a compromise that tries best to take the advantages of nix, but being conservative so that we can never fail to enter the shell?

Reply:
The problem is that unless we perfectly select the packages version from nix so they match what `cabal` will select when creating its build plan, we cannot ensure that cabal won't try to rebuild dependencies inside the shell.

Unless we know how to force `cabal` to use what's available in the package database provided by nix. I don't know how to do that.

Another solution would be to use https://github.com/input-output-hk/haskell.nix and let it parses the cabal build plan so we can be sure that we'll fetch the dependencies which will be used by cabal. This would be the perfect solution, because it will also mean that `nix build` will uses the same dependencies as the one used by `cabal build`.

Reply:
I think that given how frequently the HLS deps change (or even just the index-state), trying to keep up with things manually is doomed to failure. Something like `haskell.nix` that does it automatically can work (I use it at work), but I think is too complex for this project.

Letting cabal build the Haskell deps is dumb and a little slow, but requires almost no maintenance and tends to work.

Reply:
> Unless we know how to force cabal to use what's available in the package database provided by nix. I don't know how to do that.

Actually, @berberman, you may know more than me about that. For some reasons, `cabal build` in `nix-shell` (or `nix develop`) won't fetch/build from hackage. However, it does for `nix develop .#haskell-language-server-ghcXXX-dev` for ghc 9.0 and 9.2. For example, it fetch and build `lsp-1.4.0.0` from hackage, even if the same package with same version is available inside `ghc-pkg list`. I may have missed something here, but how do I force `cabal` to use the package that is already available in my "local nix" package database?

Reply:
One option may be to keep both options available, so users can decide to either pick the basic shell or use the advanced one which pre-populate the nix cache with all the Haskell dependencies.

At least, if the dependencies changes, dev will still be able to use the basic shell? And it will test our motivation to maintain the precise list of dependencies in nix. (I'll try to do it, contrary to what this MR may let you think, I'm highly in favor of doing as much as possible with nix, just that I exhausted the amount of time available to fix this, and I instead decided to try a simpler approach).

@berberman, @michaelpj, do you agree if I keep both approach? However I'd like to set the "basic" shell as the default one, so we get some guarantee that it will always be usable?

Reply:
Sorry know nothing about nix but if you keep two version i think a good way to keep them up to date would be add both to ci

Reply:
The one with all the deps has almost always been broken, if that's a useful data point.

I think in principle you can force cabal to use only what's in the package-db by setting `CABAL_CONFIG` to something that doesn't have any remote repositories. But then you're really out of luck if it doesn't work.

Reply:
I'm okay with having both, I just don't want to maintain the complicated one :p

Reply:
Super.

Reply:
I've pushed a change where both type of shell exists.